consensus/state_processing/src/block_replayer.rs

```diff
@@ -0,0 +1,292 @@
+        self.pre_slot_hook = Some(hook);
+        self
+    }
+
+    /// Run a function immediately after slot processing has advanced the state to the next slot.
+    ///
+    /// The hook receives the state and a bool indicating if this state corresponds to a skipped
+    /// slot (i.e. it will not have a block applied).
+    pub fn post_slot_hook(mut self, hook: PostSlotHook<'a, E, Error>) -> Self {
+        self.post_slot_hook = Some(hook);
+        self
+    }
+
+    /// Compute the state root for `slot` as efficiently as possible.
+    ///
+    /// The `blocks` should be the full list of blocks being applied and `i` should be the index of
+    /// the next block that will be applied, or `blocks.len()` if all blocks have already been
+    /// applied.
+    fn get_state_root(
+        &mut self,
+        slot: Slot,
+        blocks: &[SignedBeaconBlock<E>],
+        i: usize,
+    ) -> Result<Option<Hash256>, Error> {
+        // If we don't care about state roots then return immediately.
+        if self.state_root_strategy == StateRootStrategy::Inconsistent {
+            return Ok(Some(Hash256::zero()));
+        }
+
+        // If a state root iterator is configured, use it to find the root.
+        if let Some(ref mut state_root_iter) = self.state_root_iter {
+            let opt_root = state_root_iter
+                .take_while(|res| res.as_ref().map_or(true, |(_, s)| *s <= slot))
+                .find(|res| res.as_ref().map_or(true, |(_, s)| *s == slot))
+                .transpose()?;
+
+            if let Some((root, _)) = opt_root {
+                return Ok(Some(root));
+            }
+        }
+
+        // Otherwise try to source a root from the previous block.
+        if i > 0 {
+            if let Some(prev_block) = blocks.get(i.safe_sub(1).map_err(BlockReplayError::from)?) {
+                if prev_block.slot() == slot {
+                    return Ok(Some(prev_block.state_root()));
+                }
+            }
+        }
+
+        self.state_root_miss = true;
+        Ok(None)
+    }
+
+    /// Apply `blocks` atop `self.state`, taking care of slot processing.
+    ///
+    /// If `target_slot` is provided then the state will be advanced through to `target_slot`
+    /// after the blocks have been applied.
+    pub fn apply_blocks(
+        mut self,
+        blocks: Vec<SignedBeaconBlock<E>>,
+        target_slot: Option<Slot>,
+    ) -> Result<Self, Error> {
+        for (i, block) in blocks.iter().enumerate() {
+            if block.slot() <= self.state.slot() {
+                continue;
+            }
+
+            while self.state.slot() < block.slot() {
+                if let Some(ref mut pre_slot_hook) = self.pre_slot_hook {
+                    pre_slot_hook(&mut self.state)?;
+                }
+
+                let state_root = self.get_state_root(self.state.slot(), &blocks, i)?;
+                per_slot_processing(&mut self.state, state_root, self.spec)
```

Comment:
Perhaps capturing the `Option<EpochProcessingSummary>` would be useful for the sort of things @macladson is doing.

Reply:
Ah yes! Very good idea! I'll store them in a `HashMap<Epoch, EpochProcessingSummary>` on the state.

Might also add a config option that avoids storing them.

Reply:
I ended up feeding them to the `post_slot_hook`, which can take care of storing them if it desires (or better yet: aggregate them as it receives them)